File: datasets/Python/Gemini/Algorithm/hard1.py
import sys

def matrixRotation(matrix, m, n, r):
    """
    Rotates a 2D matrix layer by layer in an anti-clockwise direction.

    Args:
        matrix: A list of lists representing the m x n matrix.
        m: The number of rows.
        n: The number of columns.
        r: The number of anti-clockwise rotations.
    """
    num_layers = min(m, n) // 2

    for i in range(num_layers):
        # Define boundaries for the current layer
        top, bottom = i, m - 1 - i
        left, right = i, n - 1 - i

        # A layer that is just a line has 0 perimeter, so skip
        if left >= right or top >= bottom:
            break

        # 1. Extract layer elements into a 1D list (anti-clockwise)
        layer_1d = []
        # Move down the left column
        for row in range(top, bottom + 1):
            layer_1d.append(matrix[row][left])
        # Move right across the bottom row
        for col in range(left + 1, right + 1):
            layer_1d.append(matrix[bottom][col])
        # Move up the right column
        for row in range(bottom - 1, top - 1, -1):
            layer_1d.append(matrix[row][right])
        # Move left across the top row
        for col in range(right - 1, left, -1):
            layer_1d.append(matrix[top][col])
            
        # 2. Calculate effective rotations
        perimeter = len(layer_1d)
        effective_rotations = r % perimeter
        
        # If no effective rotation needed, continue
        if effective_rotations == 0:
            continue

        # 3. Rotate the 1D list (right shift for anti-clockwise matrix rotation)
        rotated_layer = layer_1d[-effective_rotations:] + layer_1d[:-effective_rotations]
        
        # 4. Place rotated elements back into the matrix
        idx = 0
        # Move down the left column
        for row in range(top, bottom + 1):
            matrix[row][left] = rotated_layer[idx]
            idx += 1
        # Move right across the bottom row
        for col in range(left + 1, right + 1):
            matrix[bottom][col] = rotated_layer[idx]
            idx += 1
        # Move up the right column
        for row in range(bottom - 1, top - 1, -1):
            matrix[row][right] = rotated_layer[idx]
            idx += 1
        # Move left across the top row
        for col in range(right - 1, left, -1):
            matrix[top][col] = rotated_layer[idx]
            idx += 1
            
    # Print the final matrix
    for row in matrix:
        print(*row)

if __name__ == '__main__':
    # Reading input from stdin
    m, n, r = map(int, sys.stdin.readline().rstrip().split())
    matrix = []
    for _ in range(m):
        matrix.append(list(map(int, sys.stdin.readline().rstrip().split())))
    
    matrixRotation(matrix, m, n, r)

File: datasets/C++/Gemini/Algorithm/hard1.cpp
#include <iostream>
#include <vector>
#include <algorithm>

void matrixRotation(std::vector<std::vector<int>>& matrix, int m, int n, int r) {
    int num_layers = std::min(m, n) / 2;

    for (int i = 0; i < num_layers; ++i) {
        // Define boundaries for the current layer
        int top = i, bottom = m - 1 - i;
        int left = i, right = n - 1 - i;

        // A layer that is just a line has 0 perimeter, so skip
        if (left >= right || top >= bottom) {
            break;
        }

        // 1. Extract layer elements into a 1D vector (anti-clockwise)
        std::vector<int> layer_1d;
        // Move down the left column
        for (int row = top; row <= bottom; ++row) layer_1d.push_back(matrix[row][left]);
        // Move right across the bottom row
        for (int col = left + 1; col <= right; ++col) layer_1d.push_back(matrix[bottom][col]);
        // Move up the right column
        for (int row = bottom - 1; row >= top; --row) layer_1d.push_back(matrix[row][right]);
        // Move left across the top row
        for (int col = right - 1; col > left; --col) layer_1d.push_back(matrix[top][col]);
        
        // 2. Calculate effective rotations
        int perimeter = layer_1d.size();
        int effective_rotations = r % perimeter;

        // If no effective rotation needed, continue
        if (effective_rotations == 0) {
            continue;
        }

        // 3. Rotate the 1D vector. A right shift by k is a left shift by (size - k).
        // std::rotate performs a left shift.
        std::rotate(layer_1d.begin(), layer_1d.begin() + (perimeter - effective_rotations), layer_1d.end());
        
        // 4. Place rotated elements back into the matrix
        int idx = 0;
        // Move down the left column
        for (int row = top; row <= bottom; ++row) matrix[row][left] = layer_1d[idx++];
        // Move right across the bottom row
        for (int col = left + 1; col <= right; ++col) matrix[bottom][col] = layer_1d[idx++];
        // Move up the right column
        for (int row = bottom - 1; row >= top; --row) matrix[row][right] = layer_1d[idx++];
        // Move left across the top row
        for (int col = right - 1; col > left; --col) matrix[top][col] = layer_1d[idx++];
    }

    // Print the final matrix
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            std::cout << matrix[i][j] << (j == n - 1 ? "" : " ");
        }
        std::cout << std::endl;
    }
}

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    int m, n, r;
    std::cin >> m >> n >> r;

    std::vector<std::vector<int>> matrix(m, std::vector<int>(n));
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            std::cin >> matrix[i][j];
        }
    }

    matrixRotation(matrix, m, n, r);

    return 0;
}